<center><b><H1> DS 203: Assignment 4 </H1></b></center>
<center><b><H2> Vinit Awale </H1></b></center>
<center><b><H2> 18D070067 </H1></b></center>

<b><i><H3> Question 1: </H3></i></b>
Download, read, and display the data at the following URLs in in python using Google CoLab and pandas, and print the type of each variable. Comment on the difference between python data types (float, int, object
etc.) and the data types taught in class (categorical/nominal, ordinal, numerical (integer, quantized,
continuous) etc.). 

1. <href>https://data.gov.in/resources/quarterly-range-wise-performance-public-facilities-operation-minor-no-or-local-937</href>
2. <href>https://finance.yahoo.com/quote/LLY/history?p=LLY </href>
3. <href> https://archive.ics.uci.edu/ml/datasets/Amphibians </href>
 

I have downloaded these datasets and saved them with the following names:
1. operation_minor.csv
2. LLY.csv
3. dataset.csv

### UPLOADING DATASET

In [ ]:
from google.colab import files
uploaded = files.upload()

Now, we have uploaded the datasets to our google colab.

### IMPORTS

In [1]:
import numpy as np
import pandas as pd

<center><H3><b>DATASET 1: OPERATION MINOR</b></H3></center>

### READING DATASET

In [2]:
dataset_1 = pd.read_csv('operation_minor.csv')

### DISPLAYING DATASET

In [3]:
dataset_1.head()

District Facility Type  Total No. of Facilities #  \
0  Alipurduar            DH                          2   
1  Alipurduar           SDH                          2   
2     Bankura            DH                          4   
3     Bankura           CHC                         23   
4     Birbhum            DH                          2   

   No. of facilities reporting nil performance *  \
0                                              1   
1                                              1   
2                                              1   
3                                              9   
4                                              0   

   Performance - Overall Average **  Performance - Maximum  \
0                               274                    274   
1                                64                     64   
2                               302                    487   
3                               179                    607   
4                               561                    589   

   Performance - Minimum  No. of facilities by performance - 1 to 30  \
0                    274                                           0   
1                     64                                           0   
2                     93                                           0   
3                      6                                           2   
4                    534                                           0   

   No. of facilities by performance - 31 to 150  \
0                                             0   
1                                             1   
2                                             1   
3                                             4   
4                                             0   

   No. of facilities by performance - 151 to 300  \
0                                              1   
1                                              0   
2                                              0   
3                                              6   
4                                              0   

   No. of facilities by performance - >300  
0                                        0  
1                                        0  
2                                        2  
3                                        2  
4                                        2

In [4]:
dataset_1.tail()

District Facility Type  Total No. of Facilities #  \
69  South Twenty Four Parganas           PHC                         79   
70              Uttar Dinajpur            DH                          1   
71              Uttar Dinajpur           SDH                          2   
72              Uttar Dinajpur           CHC                         10   
73              Uttar Dinajpur           PHC                         24   

    No. of facilities reporting nil performance *  \
69                                             77   
70                                              0   
71                                              1   
72                                              8   
73                                             23   

    Performance - Overall Average **  Performance - Maximum  \
69                                39                     58   
70                               805                    805   
71                              1028                   1028   
72                                82                    144   
73                                46                     46   

    Performance - Minimum  No. of facilities by performance - 1 to 30  \
69                     20                                           1   
70                    805                                           0   
71                   1028                                           0   
72                     20                                           1   
73                     46                                           0   

    No. of facilities by performance - 31 to 150  \
69                                             1   
70                                             0   
71                                             0   
72                                             1   
73                                             1   

    No. of facilities by performance - 151 to 300  \
69                                              0   
70                                              0   
71                                              0   
72                                              0   
73                                              0   

    No. of facilities by performance - >300  
69                                        0  
70                                        1  
71                                        1  
72                                        0  
73                                        0

In [5]:
dataset_1.shape

(74, 11)

Hence, we can see that there are 11 columns (features) and 74 rows (observations) in the dataset.

### DATA TYPES OF VARIABLES IN DATASET

In [6]:
dataset_1.dtypes

District                                         object
Facility Type                                    object
Total No. of Facilities #                         int64
No. of facilities reporting nil performance *     int64
Performance - Overall Average **                  int64
Performance - Maximum                             int64
Performance - Minimum                             int64
No. of facilities by performance - 1 to 30        int64
No. of facilities by performance - 31 to 150      int64
No. of facilities by performance - 151 to 300     int64
No. of facilities by performance - >300           int64
dtype: object

<center><H3><b>DATASET 2: LLY</b></H3></center>

### READING DATASET

In [7]:
dataset_2 = pd.read_csv('LLY.csv')

### DISPLAYING DATASET

In [8]:
dataset_2.head()

Date        Open        High         Low       Close   Adj Close  \
0  2020-09-04  148.770004  152.580002  147.949997  150.910004  148.377274   
1  2020-09-08  150.949997  151.309998  148.500000  149.309998  146.804123   
2  2020-09-09  150.619995  154.119995  150.309998  151.970001  149.419495   
3  2020-09-10  149.880005  150.600006  145.720001  146.220001  143.765991   
4  2020-09-11  147.559998  149.190002  146.880005  148.089996  145.604614   

    Volume  
0  4361800  
1  4351200  
2  2328500  
3  3961300  
4  3001400

In [9]:
dataset_2.tail()

Date        Open        High         Low       Close   Adj Close  \
247  2021-08-30  261.029999  264.070007  259.609985  262.309998  262.309998   
248  2021-08-31  262.839996  263.640015  257.859985  258.290009  258.290009   
249  2021-09-01  258.540009  258.769989  254.490005  257.399994  257.399994   
250  2021-09-02  257.410004  259.730011  256.470001  259.570007  259.570007   
251  2021-09-03  260.829987  260.989990  256.779999  259.529999  259.529999   

      Volume  
247  1300700  
248  2733400  
249  2956800  
250  1469700  
251  1124500

In [10]:
dataset_2.shape

(252, 7)

Hence, we can see that there are 7 columns (features) and 252 rows (observations) in the dataset.

### DATA TYPES OF VARIABLES IN DATASET

In [11]:
dataset_2.dtypes

Date          object
Open         float64
High         float64
Low          float64
Close        float64
Adj Close    float64
Volume         int64
dtype: object

<center><H3><b>DATASET 3</b></H3></center>

### READING DATASET

In [12]:
dataset_3 = pd.read_csv('dataset.csv')

### DISPLAYING DATASET

In [13]:
dataset_3.head()

Integer;Categorical;Numerical;Numerical;Categorical;Categorical;Categorical;Categorical;Categorical;Categorical;Categorical;Numerical;Ordinal;Ordinal;Categorical;Categorical;Label 1;Label 2;Label 3;Label 4;Label 5;Label 6;Label 7
0  ID;Motorway;SR;NR;TR;VR;SUR1;SUR2;SUR3;UR;FR;O...                                                                                                                                                                                   
1  1;A1;600;1;1;4;6;2;10;0;0;50;0;0;0;1;0;0;0;0;0...                                                                                                                                                                                   
2  2;A1;700;1;5;1;10;6;10;3;1;75;1;1;0;1;0;1;1;0;...                                                                                                                                                                                   
3  3;A1;200;1;5;1;10;6;10;3;4;75;1;1;0;1;0;1;1;0;...                                                                                                                                                                                   
4  4;A1;300;1;5;0;6;10;2;3;4;25;0;0;0;1;0;0;1;0;0...

Hence, we can see that the dataset in not in the standard format. Hence, we have to convert it to the standard format. Also we can figure out that the features of the dataset are as follows:  
['ID','Motorway','SR','NR','TR','VR','SUR1','SUR2','SUR3','UR','FR','OR','RR','BR','MR','CR','Green frogs','Brown frogs','Common toad','Fire-bellied toad','Tree frog','Common newt','Great crested newt']

In [14]:
features = ['ID','Motorway','SR','NR','TR','VR','SUR1','SUR2','SUR3','UR','FR','OR','RR','BR','MR','CR','Green frogs','Brown frogs','Common toad','Fire-bellied toad','Tree frog','Common newt','Great crested newt']

In [15]:
dataset_3.shape

(190, 1)

From the shape of the dataset, we can figure out that this non standard datset has only one column and it is named as:  
 "Integer;Categorical;Numerical;Numerical;Categorical;Categorical;Categorical;Categorical;Categorical;Categorical;Categorical;Numerical;Ordinal;Ordinal;Categorical;Categorical;Label 1;Label 2;Label 3;Label 4;Label 5;Label 6;Label 7"

In [16]:
# Store the name of the non-standard feature in a variable

non_standard_feature =  "Integer;Categorical;Numerical;Numerical;Categorical;Categorical;Categorical;Categorical;Categorical;Categorical;Categorical;Numerical;Ordinal;Ordinal;Categorical;Categorical;Label 1;Label 2;Label 3;Label 4;Label 5;Label 6;Label 7"   

### Printing the first two elements of the dataset

In [17]:
dataset_3[non_standard_feature][0:2]

0    ID;Motorway;SR;NR;TR;VR;SUR1;SUR2;SUR3;UR;FR;O...
1    1;A1;600;1;1;4;6;2;10;0;0;50;0;0;0;1;0;0;0;0;0...
Name: Integer;Categorical;Numerical;Numerical;Categorical;Categorical;Categorical;Categorical;Categorical;Categorical;Categorical;Numerical;Ordinal;Ordinal;Categorical;Categorical;Label 1;Label 2;Label 3;Label 4;Label 5;Label 6;Label 7, dtype: object

#### Now let us convert the dataframe into a standrd format.

In [18]:
# Make a empty dataframe
dataset_3_standard = pd.DataFrame()

In [19]:
dataset_3_standard[features] = np.zeros(len(features))      # Add the features to the empty dataframe

In [20]:
dataset_3_standard                                          # Print the empty dataframe

Empty DataFrame
Columns: [ID, Motorway, SR, NR, TR, VR, SUR1, SUR2, SUR3, UR, FR, OR, RR, BR, MR, CR, Green frogs, Brown frogs, Common toad, Fire-bellied toad, Tree frog, Common newt, Great crested newt]
Index: []

[0 rows x 23 columns]

In [21]:
## Add the obseravations to the dataframe in a stanard format

for i in range(1, dataset_3.shape[0]):
    dataset_3_standard.loc[len(dataset_3_standard.index)] = dataset_3[non_standard_feature][i].split(';')

In [22]:
dataset_3_standard

ID Motorway    SR NR  TR VR SUR1 SUR2 SUR3 UR  ... BR MR CR Green frogs  \
0      1       A1   600  1   1  4    6    2   10  0  ...  0  0  1           0   
1      2       A1   700  1   5  1   10    6   10  3  ...  1  0  1           0   
2      3       A1   200  1   5  1   10    6   10  3  ...  1  0  1           0   
3      4       A1   300  1   5  0    6   10    2  3  ...  0  0  1           0   
4      5       A1   600  2   1  4   10    2    6  0  ...  5  0  1           0   
..   ...      ...   ... ..  .. ..  ...  ...  ... ..  ... .. .. ..         ...   
184  185      S52  2300  1  12  3    2    2    1  0  ...  1  0  1           0   
185  186      S52   300  1  14  2    7   10    2  0  ...  5  0  1           1   
186  187      S52   500  1   1  4    1   10    2  0  ...  5  0  1           1   
187  188      S52   300  1  12  3    2    1    6  0  ...  0  0  1           0   
188  189      S52   300  1  12  3    2    6   10  0  ...  1  0  1           0   

    Brown frogs Common toad Fire-bellied toad Tree frog Common newt  \
0             0           0                 0         0           0   
1             1           1                 0         0           1   
2             1           1                 0         0           1   
3             0           1                 0         0           0   
4             1           1                 1         0           1   
..          ...         ...               ...       ...         ...   
184           1           0                 0         0           0   
185           1           1                 1         0           1   
186           1           1                 1         0           1   
187           1           1                 0         0           0   
188           1           1                 0         0           0   

    Great crested newt  
0                    0  
1                    0  
2                    0  
3                    0  
4                    1  
..                 ...  
184                  0  
185                  0  
186                  0  
187                  0  
188                  0  

[189 rows x 23 columns]

In [23]:
dataset_3_standard.shape

(189, 23)

Hence, we can see that there are 23 columns (features) and 190 rows (observations) in the dataset.

### DATA TYPES OF VARIABLES IN DATASET

In [24]:
dataset_3_standard.dtypes

ID                    object
Motorway              object
SR                    object
NR                    object
TR                    object
VR                    object
SUR1                  object
SUR2                  object
SUR3                  object
UR                    object
FR                    object
OR                    object
RR                    object
BR                    object
MR                    object
CR                    object
Green frogs           object
Brown frogs           object
Common toad           object
Fire-bellied toad     object
Tree frog             object
Common newt           object
Great crested newt    object
dtype: object